README.md
[![](https://github.com/wutsi/sdui/actions/workflows/master.yml/badge.svg)](https://github.com/wutsi/sdui/actions/workflows/master.yml)

[![pub package](https://img.shields.io/pub/v/sdui.svg)](https://pub.dev/packages/sdui)
![](https://img.shields.io/badge/licence-MIT-yellow.svg)
![](https://img.shields.io/badge/language-flutter_3.0-darkblue.svg)
![](https://img.shields.io/badge/language-dart_2.17-darkblue.svg)

# SDUI

*SDUI* make it easy to implement Server Driven UI pattern on flutter.

- The server decides what to render by describing in a JSON the widgets to render.
- The Flutter screen parse the JSON and build the widgets

Kind like HTML... but not really.

## Example

### The Server

Here is an example of JSON returned by the URL `POST http://myapp.herokuapp.com/screens/profile`:

```json
{
    "type": "Screen",
    "appBar": {
        "type": "AppBar",
        "attributes": {
            "title": "Profile"
        }
    },
    "child": {
        "type": "Form",
        "attributes": {
            "padding": 10
        },
        "children": [
            {
                "type": "Input",
                "attributes": {
                    "name": "first_name",
                    "value": "Ray",
                    "caption": "First Name",
                    "maxLength": 30
                }
            },
            {
                "type": "Input",
                "attributes": {
                    "name": "last_name",
                    "value": "Sponsible",
                    "caption": "Last Name",
                    "maxLength": 30
                }
            },
            {
                "type": "Input",
                "attributes": {
                    "name": "email",
                    "value": "ray.sponsible@gmail.com",
                    "caption": "Email *",
                    "required": true
                }
            },
            {
                "type": "Input",
                "attributes": {
                    "type": "date",
                    "name": "birth_date",
                    "caption": "Date of Birth"
                }
            },
            {
                "type": "Input",
                "attributes": {
                    "type": "Submit",
                    "name": "submit",
                    "caption": "Create Profile"
                },
                "action": {
                    "type": "Command",
                    "url": "https://myapp.herokuapp.com/commands/save-profile",
                    "prompt": {
                        "type": "Dialog",
                        "attributes": {
                            "type": "confirm",
                            "title": "Confirmation",
                            "message": "Are you sure you want to change your profile?"
                        }
                    }
                }
            }
        ]
    }
}
```

### The UI in Flutter

```dart
import 'package:flutter/material.dart';
import 'package:sdui/sdui.dart';

void main() async {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(title: 'Demo', initialRoute: '/', routes: _routes());
  }

  Map<String, WidgetBuilder> _routes() =>
      {
        '/': (context) =>
        const DynamicRoute(
            provider: HttpRouteContentProvider(
                'http://www.myapp.com/screens/profile'))
      };
}
```

### Screenshots

| Screen                            | Date Picker                        | Alert Dialog                        |
|-----------------------------------|------------------------------------|-------------------------------------|
| ![](doc/images/screenshot-01.png) | ![](doc/images/screenshot-02.png)  | ![](doc/images/screenshot-03.png)   |

## Widgets

In Flutter, UI is composed of a hierarchy of Widgets. A widget is a visual element on a screen.

*SDUI* described widgets with the following json structure:

```
{
    "type": "...",
    "attributes": {
      "padding": 12,
      "color": "#ff0000",
      ...
    },
    "children": [
      ...
    ]
}
```

- `type`: indicates the type of widget
- `attributes`: key/value pair of the attributes of the widget. Ex: `caption`, `color`, `padding`, `spacing` etc.
- `children`: The list of children widgets

### Widget Library

- Navigation widgets
    - [Dialog](https://pub.dev/documentation/sdui/latest/sdui/SDUIDialog-class.html)
    - [AppBar](https://pub.dev/documentation/sdui/latest/sdui/SDUIAppBar-class.html)
    - [Page](https://pub.dev/documentation/sdui/latest/sdui/SDUIPage-class.html)
    - [PageView](https://pub.dev/documentation/sdui/latest/sdui/SDUIPageView-class.html)
    - [Screen](https://pub.dev/documentation/sdui/latest/sdui/SDUIScreen-class.html)
- Layout widgets
    - [Center](https://pub.dev/documentation/sdui/latest/sdui/SDUICenter-class.html)
    - [Column](https://pub.dev/documentation/sdui/latest/sdui/SDUIColumn-class.html)
    - [Container](https://pub.dev/documentation/sdui/latest/sdui/SDUIContainer-class.html)
    - [Expanded](https://pub.dev/documentation/sdui/latest/sdui/SDUIExpanded-class.html)
    - [FittedBox](https://pub.dev/documentation/sdui/latest/sdui/SDUIFittedBox-class.html)
    - [Flexible](https://pub.dev/documentation/sdui/latest/sdui/SDUIFlexible-class.html)
    - [GridView](https://pub.dev/documentation/sdui/latest/sdui/SDUIGridView-class.html)
    - [Positioned](https://pub.dev/documentation/sdui/latest/sdui/SDUIPositioned-class.html)
    - [Row](https://pub.dev/documentation/sdui/latest/sdui/SDUIRow-class.html)
    - [Stack](https://pub.dev/documentation/sdui/latest/sdui/SDUIStack-class.html)
    - [SingleChildScrollView](https://pub.dev/documentation/sdui/latest/sdui/SDUISingleChildScrollView-class.html)
    - [Spacer](https://pub.dev/documentation/sdui/latest/sdui/SDUISpacer-class.html)
- Images/Icons
    - [CarouselSlider](https://pub.dev/documentation/sdui/latest/sdui/SDUICarouselSlider-class.html)
    - [CircleAvatar](https://pub.dev/documentation/sdui/latest/sdui/SDUICircleAvatar-class.html)
    - [Icon](https://pub.dev/documentation/sdui/latest/sdui/SDUIIcon-class.html)
    - [IconButton](https://pub.dev/documentation/sdui/latest/sdui/SDUIIconButton-class.html)
    - [Image](https://pub.dev/documentation/sdui/latest/sdui/SDUIImage-class.html)
    - [PhotoView](https://pub.dev/documentation/sdui/latest/sdui/SDUIPhotoView-class.html)
- Input widgets
    - [Button](https://pub.dev/documentation/sdui/latest/sdui/SDUIButton-class.html)
    - [Camera](https://pub.dev/documentation/sdui/latest/sdui/SDUICamera-class.html)
    - [DropdownButton](https://pub.dev/documentation/sdui/latest/sdui/SDUIDropdownButton-class.html)
    - [DropdownMenuItem](https://pub.dev/documentation/sdui/latest/sdui/SDUIDropdownMenuItem-class.html)
    - [Form](https://pub.dev/documentation/sdui/latest/sdui/SDUIForm-class.html)
    - [IconButton](https://pub.dev/documentation/sdui/latest/sdui/SDUIIconButton-class.html)
    - [Input](https://pub.dev/documentation/sdui/latest/sdui/SDUIInput-class.html)
    - [ListView](https://pub.dev/documentation/sdui/latest/sdui/SDUIListView-class.html)
    - [ListItem](https://pub.dev/documentation/sdui/latest/sdui/SDUIListItem-class.html)
    - [ListItemSwitch](https://pub.dev/documentation/sdui/latest/sdui/SDUIListItemSwitch-class.html)
    - [MoneyText](https://pub.dev/documentation/sdui/latest/sdui/SDUIMoneyText-class.html)
    - [MoneyWithKeyboard](https://pub.dev/documentation/sdui/latest/sdui/SDUIMoneyWithKeyboard-class.html)
    - [MoneyWithSlider](https://pub.dev/documentation/sdui/latest/sdui/MoneyWithSlider-class.html)
    - [PinWidthKeyboard](https://pub.dev/documentation/sdui/latest/sdui/SDUIPinWidthKeyboard-class.html)
    - [QrImage](https://pub.dev/documentation/sdui/latest/sdui/SDUIQrImage-class.html)
    - [QrView](https://pub.dev/documentation/sdui/latest/sdui/SDUIQrView-class.html)
    - [Radio](https://pub.dev/documentation/sdui/latest/sdui/SDUIRadio-class.html)
    - [RadioGroup](https://pub.dev/documentation/sdui/latest/sdui/SDUIRadioGroup-class.html)
    - [SearchableDropdown](https://pub.dev/documentation/sdui/latest/sdui/SDUISearchableButton-class.html)
    - [Text](https://pub.dev/documentation/sdui/latest/sdui/SDUIText-class.html)
- Other
    - [Badge](https://pub.dev/documentation/sdui/latest/sdui/SDUIBadge-class.html)
    - [Chat](https://pub.dev/documentation/sdui/latest/sdui/SDUIChat-class.html)
    - [Chip](https://pub.dev/documentation/sdui/latest/sdui/SDUIChip-class.html)
    - [ExpandablePanel](https://pub.dev/documentation/sdui/latest/sdui/SDUIExpandablePanel-class.html)
    - [Html](https://pub.dev/documentation/sdui/latest/sdui/SDUIHtml-class.html)

### Global Variables

- `sduiErrorState`: Function for building the error state.
- `sduiLoadingState`: Function for building the loading state.
- `sduiProgressIndicator`: Function for building the progress indicator.
- `sduiRouteObserver`: Route observer that reload each page on navigation.
- `sduiAnalytics`: Analytics class.
  See [SDUIAnalytics](https://pub.dev/documentation/sdui/latest/sdui/SDUIAnalytics-class.html)
- `sduiCameras`: List of available cameras. Empty by default, must be initialized the application

## Actions

With actions, you can:

- Execute a command on a server (Ex: Saver User Profile, Delete User Account etc.)
- Navigate to another screen.
- Prompt a message to user.

*SDUI* described actions with the following json structure:

```
{
    "type": "...",
    "attributes": {
      ...
    },
    ...
    "action": {
      "type": "...",
      "url": "...",
      "prompt": {
        "type": "...",
        "title": "Confirmation",
        "message": "Are you sure you want to change your profile?"
      }
    }

}
```

- `type`: Defines the type of action:
    - `Route`: To redirect to another route
    - `Page`: To redirect to another page, in the context of [PageView]
    - `Command`: Remote action to execute. The screen is associated with a URL that will execute the command, and
      redirect the user to the next screen
    - `Share`: Share a message to user via email/messenger/whatsapp etc.
    - `Navigate`: Navigate user to a web page
- `url`: is the URL associated with the action
    - `route:/..`: redirect users to previous route
    - `route:/~`: redirect users to 1st route
    - URL starting with `route:/<ROUTE_NAME>` redirect user the a named route. (Ex: ``route:/checkout``)
    - URL starting with `http://` or `https` redirect user to a server driven page
    - `page:/<PAGE_NUMBER>`: redirect users to a given page. `<PAGE_NUMBER>` is the page index (starting with `0`).
- `replacement`: For `type=Route`, this indicate if we replace the current view or navigate.
- `parameters`: Parameters to add to the URL where to redirect to
- `message`: Message to share
- `prompt`: Shows a Dialog before execute the action
    - `type`: The type of prompt (Exemple: `Confirm`, `Error`, `Warning`, `Information`)
    - `title`: Title of the alert box to open
    - `message`: Message to display to the user

## Build your own Widget

You can integrate your own widget into `sdui`.

### Step 1: Create your instance of ``SDUIWidget``

This is an example of widget that render a text with a margin and padding. The widget has the following attributes

- `text`: The text to display
- `padding`: The padding value
- `margin`: The margin value

```dart
    class MyWidget extends SDUIWidget {
  String text = '';
  double padding = 10.0;
  double margin = 10.0;

  /// This method will be called by [SDUIParser] to read the widget attributes from the JSON data
  @override
  SDUIWidget fromJson(Map<String, dynamic>? json) {
    text = json?['caption'] ?? '';
    margin = json?['margin'] ?? 10.0;
    padding = json?['padding'] ?? 10.0;
    return this;
  }

  /// This method will be called when rendering the page to create the Flutter widget
  @override
  Widget toWidget(BuildContext context) =>
      Container(
        padding: EdgeInsets.all(padding),
        margin: EdgeInsets.all(margin),
        child: Text(
          text,
          style:
          const TextStyle(fontWeight: FontWeight.bold, color: Colors.red),
        ),
      );
}
```

### Step 2: Register the widget

The Widget is registered into SDUI and associated with the type `MyWidget`

```dart
void main() async {
  // Register 3rd party widget
  SDUIWidgetRegistry.getInstance().register('MyWidget', () => MyWidget());

  runApp(constMyApp());
}

class MyApp extends StatelessWidget {
  // ...
}
```

### Step 3: Add the widget into the JSONs

Here is an example of JSON with our 3rd party widget

```json
{
    "type": "Screen",
    "appBar": {
        "type": "AppBar",
        "attributes": {
            "title": "Home"
        }
    },
    "child": {
        "type": "Center",
        "children": [
            {
                "type": "MyWidget",
                "attributes": {
                    "caption": "3rd Party Widget",
                    "padding": 5.0,
                    "margin": 5.0
                }
            }
        ]
    }
}
```


CHANGELOG.md
# CHANGELOG

## [0.1.200] 2023-02-03

## Changes

- upgrade to syncfusion_flutter_charts v21.1.41
- upgrade to badges v3.1.1

## [0.1.200] 2023-02-03

## Changes

- upgrade to gradle v8
- upgrade to kotlin v1.8.20
- upgrade to syncfusion_flutter_charts v21.0.0
- upgrade to badges v3.0.0

## [0.1.199] 2023-02-03

## Changes

- Add local to MoneyText

## [0.1.198] 2023-01-19

## Changes

- Add support for more chart

## [0.1.197] 2023-01-14

## Changes

- Log error when downloading images

## Fixes

- Fix the `ListItemSwitch` state on selection
- French translations for time selector

## [0.1.194] 2023-01-04

- FIX: Reset date/time default value

## [0.1.193] 2023-01-03

- FIX: Busy state of ListView switch
- CHANGE: Reset date/time default value

## [0.1.192] 2022-12-25

- CHANGE: Add alignment to MoneyText
- CHANGE: Use `SDUINull` for unsupported widgets

## [0.1.190] 2022-12-24

- ADD: Add chart component

## [0.1.183] 2022-12-24

- CHANGE: Update dependencies

## [0.1.181] 2022-12-21

- CHANGE: Handle response after uploading a file

## [0.1.180] 2022-11-22

- ADD: Add ClipRRect

## [0.1.179] 2022-11-20

- CHANGE: Remove HTML component
- CHANGE: Update dependencies
- CHANGE: Cleanup logs

## [0.1.175] 2022-11-18

- FIX: Fix icon of SwitchListTile

## [0.1.174] 2022-11-18

- CHANGE: Add foregroundColor to AvararCircle

## [0.1.173] 2022-10-09

- CHANGE: Add localized text

## [0.1.171] 2022-10-09

- CHANGE: Initialize the chat text

## [0.1.170] 2022-10-08

- CHANGE: Add message handle to handle FCM local message
- CHANGE: Send receive notification on loaded messages

## [0.1.162] 2022-10-07

- CHANGE: Improve error handling

## [0.1.161] 2022-10-07

- ADD: Link preview

## [0.1.160] 2022-10-07

- FIX: https://github.com/lcuis/search_choices/issues/91

## [0.1.158] 2022-10-06

- CHANGE: Add `tenantId` and `deviceId`
- CHANGE: Update dependencies
- CHANGE: Disable `search_choices` because of https://github.com/lcuis/search_choices/issues/91

## [0.1.153] 2022-10-03

- CHANGE: Handle RTM events

## [0.1.150] 2022-10-03

- CHANGE: Add support for Chat Websocket RTM API (Real Time Messaging)

## [0.1.148] 2022-10-03

- CHANGE: Improve chat UI

## [0.1.146] 2022-10-01

- ADD: Chat component
- CHANGE: Remove button padding when not provided
- CHANGE: Add font size to bottom bar

## [0.1.141] 2022-09-22

- CHANGE: Handle opening app from push notification

## [0.1.128] 2022-09-20

- CHANGE: Integrate firebase messaging with iOS

## [0.1.125] 2022-09-18

- CHANGE: Integrate firebase messaging with Android

## [0.1.118] 2022-09-12

- CHANGE: remove `OverlaySupport`

## [0.1.117] 2022-09-12

- CHANGE: Wrap `Scaffold` and `DefaultTabController` in `OverlaySupport` so that we can display push notification

## [0.1.116] 2022-09-10

- CHANGE: Update the font size of PinWithKeyboard

## [0.1.114] 2022-09-08

- CHANGE: Use cached images for Icons

## [0.1.113] 2022-09-01

- FIX: Handle initial deep-links only once

## [0.1.111] 2022-08-31

- CHANGE: Fix the state of ListItemSwitch

## [0.1.107] 2022-08-28

- CHANGE: Improve button action handling

## [0.1.106] 2022-08-26

- CHANGE: Improve logs
- CHANGE: Set the key for Default

## [0.1.103] 2022-08-26

- CHANGE: Fix execution of ListItemSwitch

## [0.1.102] 2022-08-25

- CHANGE: add `inputFormatterRegex` to `Input`

## [0.1.101] 2022-08-25

- FIX: SwitchListTile - disable the control when executing the action

## [0.1.100] 2022-08-19

- CHANGE: Add support for deeplinks

## [0.1.98] 2022-08-19

- CHANGE: Improve navigation to home page

## [0.1.97] 2022-08-16

- CHANGE: Maps id to key on
    - NumericKeyboard
    - PageView

## [0.1.96] 2022-08-16

- CHANGE: Maps id to key on
    - Badge
    - Chip
    - Column
    - Container
    - CircleAvatar
    - Button
    - Dropdown
    - Image
    - Input
    - ListView
    - ListItem
    - Money
    - MoneyWithKeyboard
    - MoneyWithSlide
    - PinWidthKeyboard
    - Radio
    - RadioGroup
    - SearchableDropdown
    - Row
    - Screen
    - Text

## [0.1.95] 2022-08-13

- CHANGE: Add more logs when redirecting to /

## [0.1.94] 2022-08-13

- CHANGE: Set the PIN text color

## [0.1.91] 2022-08-12

- CHANGE: Downgrade url_launcher from 6.5.1

## [0.1.89] 2022-08-12

- FIX: Increase the dialog content max-lines from 3 to 10 - to prevent dialog content to be cut-off
- CHANGE: Upgrade Android gradle
- CHANGE: Downgrade file_picker from 5.0.0 to 4.5.1 - to fix iOS build issues

## [0.1.88] 2022-07-26

- CHANGE: Upgrade to file_picker: ^5.0.0

## [0.1.87] 2022-07-23

- FIX: PageView scroll physics set to none by default

## [0.1.86] 2022-07-23

- FIX: Fix validation of email, url
- FIX: Fix keyboard type for text fields
- FIX: Trip text field values

## [0.1.85] 2022-07-04

- ADD: Html component

## [0.1.84] 2022-07-04

- ADD: FittedBox

## [0.1.83] 2022-06-28

- CHANGE: Upgrade to pinput v2.0.2

## [0.1.82] 2022-06-20

- CHANGE: Show uploaded filename

## [0.1.80] 2022-06-18

- FIX: Update SDK version

## [0.1.78] 2022-06-17

- ADD: Timeout component
- CHANGE: Set text capitalization in Input field

## [0.1.70] 2022-05-22

- FIX: In dialog box, make buttons larger
- CHANGE: Set HTTP timeout to 60 secs

## [0.1.69] 2022-05-22

- FIX: State of dropdown on selection

## [0.1.68] 2022-05-15

- FIX: Initialization of phone number input
- ADD: Add `initialCountry` for setting the initial flag of phone number

## [0.1.67] 2022-05-10

- UPDATE: route:~ redirect to home page

## [0.1.66] 2022-05-02

- UPDATE: Return string on upload

## [0.1.65] 2022-04-25

- ADD: MoneyText not bold by default

## [0.1.64] 2022-04-24

- ADD: ExpandablePanel

## [0.1.59] 2022-04-22

- CHANGE: Set initial index of tab

## [0.1.58] 2022-04-16

- CHANGE: Update component dependencies

## [0.1.57] 2022-04-16

- ADD: Track action events

## [0.1.50] 2022-04-15

- ADD: Add `SDUIWidgetRegistry` to integrate 3rd party widgets

## [0.1.49] 2022-04-11

- CHANGE: Log when no attributes found - See https://github.com/wutsi/sdui/issues/15

lib/src/chat.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:flutter_chat_types/flutter_chat_types.dart' as types;
import 'package:flutter_chat_ui/flutter_chat_ui.dart';
import 'package:logger/logger.dart';
import 'package:sdui/sdui.dart';
import 'package:uuid/uuid.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

/// Chat widget based on [flutter_chat_ui](https://pub.dev/packages/flutter_chat_ui)
class SDUIChat extends SDUIWidget {
  String? fetchMessageUrl;
  String? rtmUrl;
  String? roomId;
  String? userId;
  String? userFirstName;
  String? userLastName;
  String? userPictureUrl;
  String? recipientUserId;
  String? language;
  double? fontSize;
  String? receivedMessageBackground;
  String? receivedMessageTextColor;
  String? sentMessageBackground;
  String? sentMessageTextColor;
  bool? showUserNames;
  bool? showUserAvatars;
  String? tenantId;
  String? deviceId;
  String? text;

  @override
  SDUIWidget fromJson(Map<String, dynamic>? json) {
    rtmUrl = json?["rtmUrl"];
    fetchMessageUrl = json?["fetchMessageUrl"];
    userId = json?["userId"];
    roomId = json?["roomId"];
    userFirstName = json?["userFirstName"];
    userLastName = json?["userLastName"];
    userPictureUrl = json?["userPictureUrl"];
    recipientUserId = json?["recipientUserId"];
    language = json?["language"];
    fontSize = json?["fontSize"];
    receivedMessageBackground = json?["receivedMessageBackground"];
    receivedMessageTextColor = json?["receivedMessageTextColor"];
    sentMessageBackground = json?["sentMessageBackground"];
    sentMessageTextColor = json?["sentMessageTextColor"];
    showUserNames = json?["showUserNames"];
    showUserAvatars = json?["showUserAvatars"];
    tenantId = json?["tenantId"];
    deviceId = json?["deviceId"];
    text = json?["text"];
    return super.fromJson(json);
  }

  @override
  Widget toWidget(BuildContext context) => _ChatWidgetStateful(this);
}

class _ChatWidgetStateful extends StatefulWidget {
  final SDUIChat delegate;

  const _ChatWidgetStateful(this.delegate, {Key? key}) : super(key: key);

  @override
  // ignore: no_logic_in_create_state
  State<StatefulWidget> createState() => _ChatWidgetState(delegate);
}

class _ChatWidgetState extends State<_ChatWidgetStateful> {
  final Logger _logger = LoggerFactory.create('_ChatWidgetState');
  final SDUIChat _delegate;
  final _textEditingController = InputTextFieldController();

  List<types.Message> _messages = [];
  types.User _user = const types.User(id: '');
  RTM? _rtm;
  int _page = 0;

  _ChatWidgetState(this._delegate);

  @override
  void initState() {
    super.initState();

    _textEditingController.text = (_delegate.text ?? "");

    // Current user
    _user = types.User(
        id: _delegate.userId ?? "",
        firstName: _delegate.userFirstName,
        lastName: _delegate.userLastName,
        imageUrl: _delegate.userPictureUrl);

    // Fetch messages
    _fetchMessages(0);

    // Connect to RTM API
    if (_delegate.rtmUrl != null) {
      _rtm = RTM(
          roomId: _delegate.roomId ?? const Uuid().toString(),
          userId: _delegate.userId ?? '',
          url: _delegate.rtmUrl!,
          messageHandler: (message) => _handleRTMMessage(message));
    }
  }

  @override
  void dispose() {
    _rtm?.bye();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) => Chat(
        showUserAvatars: _delegate.showUserAvatars ?? true,
        showUserNames: _delegate.showUserNames ?? true,
        messages: _messages,
        user: _user,
        usePreviewData: true,
        l10n: _toL10(),
        inputOptions:
            InputOptions(textEditingController: _textEditingController),
        theme: DefaultChatTheme(
          primaryColor: _primaryColor(),
          secondaryColor: _secondaryColor(),
          dateDividerMargin: const EdgeInsets.only(bottom: 5.0, top: 5),
          dateDividerTextStyle:
              TextStyle(fontSize: _fontSize(), fontWeight: FontWeight.bold),
          errorColor: Colors.red,
          inputPadding: const EdgeInsets.all(5.0),
          inputMargin: const EdgeInsets.all(1.0),
          inputBorderRadius: BorderRadius.zero,
          inputTextStyle: TextStyle(
              fontFamily: null,
              fontWeight: FontWeight.normal,
              fontSize: _fontSize()),
          messageBorderRadius: 5.0,
          messageInsetsHorizontal: 5.0,
          messageInsetsVertical: 5.0,
          receivedEmojiMessageTextStyle: TextStyle(fontSize: _fontSize()),
          receivedMessageBodyTextStyle: TextStyle(
            color: _delegate.toColor(_delegate.receivedMessageTextColor) ??
                Colors.black,
            fontSize: _fontSize(),
          ),
          sentEmojiMessageTextStyle: TextStyle(fontSize: _fontSize()),
          sentMessageBodyTextStyle: TextStyle(
            color: _delegate.toColor(_delegate.sentMessageTextColor) ??
                Colors.white,
            fontSize: _fontSize(),
          ),
        ),
        onSendPressed: (message) => _onSend(message),
        onEndReached: () => _fetchMessages(_page + 1),
        onPreviewDataFetched: (msg, preview) =>
            _onPreviewDataFetched(msg, preview),
      );

  double _fontSize() => _delegate.fontSize ?? 12.0;

  Color _primaryColor() =>
      _delegate.toColor(_delegate.sentMessageBackground) ??
      Color(int.parse('FF1D7EDF', radix: 16));

  Color _secondaryColor() =>
      _delegate.toColor(_delegate.receivedMessageBackground) ??
      Color(int.parse('FFe4edf7', radix: 16));

  ChatL10n _toL10() {
    switch (_delegate.language?.toLowerCase()) {
      case 'fr':
        return const ChatL10nFr();
      case 'de':
        return const ChatL10nDe();
      case 'es':
        return const ChatL10nEs();
      default:
        return const ChatL10nEn();
    }
  }

  void _onPreviewDataFetched(
      types.TextMessage message, types.PreviewData previewData) {
    final index = _messages.indexWhere((element) => element.id == message.id);
    final updatedMessage = (_messages[index] as types.TextMessage).copyWith(
      previewData: previewData,
    );

    setState(() {
      _messages[index] = updatedMessage;
    });
  }

  void _onSend(types.PartialText message) {
    _logger.i('_onSend $message');

    // Create the message
    final msg = types.TextMessage(
        author: _user,
        createdAt: DateTime.now().millisecondsSinceEpoch,
        id: const Uuid().v4(),
        text: message.text,
        roomId: _delegate.roomId,
        showStatus: true,
        status: types.Status.sending,
        metadata: {
          'recipientId': _delegate.recipientUserId,
          'tenantId': _delegate.tenantId,
          'deviceId': _delegate.deviceId
        });

    // Add on the UI
    setState(() {
      _messages.insert(0, msg);
    });

    // Send to backend
    try {
      _rtm?.send(msg).then((value) =>
          _updateStatus(msg, value ? types.Status.sent : types.Status.error));
    } catch (e) {
      _logger.e('Unable to send the message', e);
      _updateStatus(msg, types.Status.error);
    }
  }

  void _updateStatus(types.Message msg, types.Status status) {
    final index = _messages.indexWhere((element) => element.id == msg.id);
    setState(() {
      _messages[index] = _messages[index].copyWith(status: status);
    });
  }

  Future<void> _fetchMessages(int page) {
    if (_delegate.fetchMessageUrl == null) return Future.value();

    var url = _fetchUrl(page);
    _logger.i('Loading messages: $url');
    return Http.getInstance().post(url, {}).then((value) {
      final messages = (jsonDecode(value) as List)
          .map((e) => types.Message.fromJson(e as Map<String, dynamic>))
          .toList();
      if (messages.isNotEmpty) {
        // Set the messages
        setState(() {
          _page = page;
          if (_page == 0) {
            _messages = messages;
          } else {
            var messageIds = messages.map((e) => e.id);
            for (var message in messages) {
              if (!messageIds.contains(message.id)) {
                _messages.add(message);
              }
            }
          }
        });

        // Notify received
        for (var message in messages) {
          _logger.i(
              'Message fetched. id=${message.id} authorId=${message.author.id} status=${message.status}');
          if (message.author.id != _delegate.userId &&
              message.status == types.Status.delivered) {
            _rtm?.received(message);
          }
        }
      }
    });
  }

  String _fetchUrl(int page) {
    var buff = StringBuffer();
    buff.write(_delegate.fetchMessageUrl);
    buff.write(_delegate.fetchMessageUrl?.contains('?') == true ? '&' : '?');
    buff.write("page=$page");
    return buff.toString();
  }

  void _handleRTMMessage(dynamic message) {
    _logger.i('_handleRTMMessage $message');

    var json = jsonDecode(message);
    var type = json['type'];
    var msg = json["chatMessage"];
    var chatMessage = msg == null
        ? null
        : types.Message.fromJson(msg as Map<String, dynamic>);

    if (type == MessageType.send.name) {
      if (chatMessage != null) {
        _handleRTMSendMessage(chatMessage);
      }
    } else if (type == MessageType.received.name) {
      if (chatMessage != null) {
        _handleRTMReceivedMessage(message);
      }
    }
  }

  void _handleRTMSendMessage(types.Message message) {
    _logger.i('_handleRTMSendMessage messageId=${message.id}');

    // Add the message
    setState(() {
      _messages.insert(0, message);
    });

    // Received
    _rtm?.received(message);
  }

  void _handleRTMReceivedMessage(types.Message message) {
    _logger.i('_handleRTMReceivedMessage messageId=${message.id}');
    _updateStatus(message, types.Status.delivered);
  }
}

@immutable
class ChatL10nFr extends ChatL10n {
  const ChatL10nFr()
      : super(
            attachmentButtonAccessibilityLabel: 'Envoyez media',
            emptyChatPlaceholder: 'Aucun message',
            fileButtonAccessibilityLabel: 'Fichier',
            inputPlaceholder: 'Message',
            sendButtonAccessibilityLabel: 'Envoyez',
            unreadMessagesLabel: 'Messages non lus');
}

enum MessageType { hello, send, bye, received }

typedef MessageHandler = void Function(dynamic message);

class RTM {
  final int reconnectDelaySeconds = 15;
  final Logger _logger = LoggerFactory.create('RTM');
  final String roomId;
  final String userId;
  final String url;
  final MessageHandler messageHandler;
  bool _connected = false;
  WebSocketChannel? _channel;
  Timer? _timer;

  RTM(
      {required this.roomId,
      required this.userId,
      required this.url,
      required this.messageHandler}) {
    // Connect
    _connect();

    // Reconnect if needed every 30 seconds
    _timer = Timer.periodic(
        Duration(seconds: reconnectDelaySeconds), (timer) => _reconnect());
  }

  void hello() {
    var data = {
      'type': MessageType.hello.name,
      'roomId': roomId,
      'userId': userId
    };
    _logger.i('hello - $data');

    _channel?.sink.add(jsonEncode(data));
  }

  Future<bool> send(types.Message message) async {
    var data = {
      'type': MessageType.send.name,
      'roomId': roomId,
      'userId': userId,
      'chatMessage': message.toJson()
    };
    _logger.i('send - $data');

    // Reconnect if needed
    if ((await _waitForReconnection()) == false) {
      _logger.i('Not connected to server');
      return false;
    }

    // Send message
    _channel?.sink.add(jsonEncode(data));
    return true;
  }

  void bye() {
    // Send message
    var data = {
      'type': MessageType.bye.name,
      'roomId': roomId,
      'userId': userId
    };
    _logger.i('bye - $data');

    // Send message
    _channel?.sink.add(jsonEncode(data));

    // Stop the timer and channel
    _timer?.cancel();
    _channel?.sink.close();
  }

  Future<bool> received(types.Message message) async {
    var data = {
      'type': MessageType.received.name,
      'roomId': roomId,
      'userId': userId,
      'chatMessage': message.toJson()
    };
    _logger.i('received - $data');

    // Reconnect if needed
    if (_connected) {
      _logger.i('Not connected to server');
      return false;
    }

    _channel?.sink.add(jsonEncode(data));
    return true;
  }

  void _onError(error) {
    _logger.i('onError $error');
  }

  void _onDone() {
    _logger.i('Server disconnected');
    _connected = false;
  }

  void _connect() {
    // Connect
    _logger.i('Connecting to $url');
    _channel = WebSocketChannel.connect(Uri.parse(url));
    _channel?.stream.listen((message) => messageHandler(message),
        onDone: () => _onDone(), onError: (error) => _onError(error));
    _connected = true;

    // Hello
    hello();
  }

  void _reconnect() {
    if (!_connected) {
      _connect();
    }
  }

  Future<bool> _waitForReconnection() async {
    if (_connected) return true;

    for (var i = 0; i < 2; i++) {
      _logger.i('...Waiting for reconnection');
      await Future.delayed(Duration(seconds: reconnectDelaySeconds));
      if (_connected) return true;
    }
    return _connected;
  }
}
